<h1>AutoGraph reference</h1>
<p><a href="index.md">Index</a></p>
<h2>Common AutoGraph errors</h2>
<h3>"WARNING: AutoGraph could not transform <code>&lt;name&gt;</code>"</h3>
<p>This warning is output when AutoGraph could not convert a function, for an
unexpected reason. The error message contains the reason why the function could
not be converted, as well as guidance on how to proceed next.</p>
<p>The exact error message may vary from version to version but in general, the
cause of the failure appears somewhere in the text, for example as
"Cause: could not get source code" or "Original error: could not get source
code".</p>
<p>Note: AutoGraph does not always output a warning. For example, constructors
are silently called without conversion.</p>
<p>When this warning is printed, the code returned by AutoGraph still executes, but
the functions indicated in the warning will be executed as they are, without
conversion. If the functions contain pure Python or graph code (for example,
they have no Tensor-dependent control flow), then the code is likely to still
run without error. However, if it contains any constructs that are only
supported in AutoGraph, expect subsequent exceptions.</p>
<p>Note: the warning is output to the <a href="https://github.com/abseil/abseil-py">abseil</a>
logger, with <code>WARNING</code> severity. To direct these warnings to <code>stdout</code>, use
<code>tf.autograph.set_verbosity(0, True)</code>.</p>
<h3>"Unable to locate the source code" or "Source not found" errors</h3>
<p>Newer versions of AutoGraph raise a <code>ConversionError</code>. Older versions print a
warning. In both cases, a similar message about finding the source code is
included.</p>
<p>These errors are raised when AutoGraph is unable to find the source code of
functions it needs to transform. See <a href="limitations.md">Limitations</a> for more
details.</p>
<h3>"WARNING: Large unrolled loop detected"</h3>
<p>This warning is output when AutoGraph detects a <code>for</code> or <code>while</code> loop that
creates TensorFlow ops and which has a large number of iterations and creates.</p>
<p>This usually indicates a loop that was intended to run as a <code>tf.while_loop</code>, but
instead runs as a Python loop.</p>
<p>For example, a training loop might mistakenly iterate over a Python <code>range</code>,
instead of <code>tf.range</code>:</p>
<p><code>num_steps = 10000
step = tf.constant(0)
for i in range(num_steps):
  step += 1
  train_step(model)</code></p>
<p>Another example is when using custom generators which AutoGraph does not
support, even if they wrap over supported iterators like Datasets:</p>
<p>```
def my_iterator(ds):
  for data in ds:
    yield data</p>
<h1>Custom iterators always dispatch to a Python for loop.</h1>
<p>for x in my_iterator(tf.data.Dataset.range(10)):
  tf.print(x)
```</p>
<p>Note: This verification is only performed when <code>__debug__</code> is <code>True</code>.</p>
<p>Note: the warning is output to the <a href="https://github.com/abseil/abseil-py">abseil</a>
logger, with <code>WARNING</code> severity. To direct these warnings to <code>stdout</code>, use
<code>tf.autograph.set_verbosity(0, True)</code>.</p>
<h3>"OperatorNotAllowedInGraphError: using a <code>tf.Tensor</code> as a Python <code>bool</code>"</h3>
<p>This exception is raised whenever a <code>tf.Tensor</code> is type-cast as a Python <code>bool</code>,
in a context where eager execution is not active. The exception is only raised
when graph execution is active, for example inside a <code>@tf.function</code> with
AutoGraph turned off.</p>
<p><strong>When AutoGraph is on</strong>, it can be caused by:
  * placing a Tensor-dependent <code>break</code>, <code>continue</code> or <code>return</code> inside a Python
    loop (see example below)
  * attempting to use a <code>tf.Tensor</code> in a list comprehension, by iterating over
    it or using it in a condition)</p>
<p>A typical example of mixing Python and TF control flow in an incompatible way
is:</p>
<p><code>for i in range(3):  # Python loop
  if i &gt; tf.constant(0):  # TF conditional
    break  # raises OperatorNotAllowedInGraphError</code></p>
<p>The way these errors are typically fixed is by ensuring all control flow is
TF control flow:</p>
<p><code>for i in tf.range(3):  # TF loop
  if i &gt; tf.constant(0):  # TF conditional
    break  # works</code></p>
<p><strong>When AutoGraph is off</strong>, it can be caused by using a <code>tf.Tensor</code> value as:</p>
<ul>
<li>the condition of an <code>if</code> or <code>while</code> statement: <code>if &lt;tensor&gt;:</code></li>
<li>the argument in a logical expression: <code>tensor and another_tensor</code></li>
<li>the argument to the <code>bool</code> built-in: <code>bool(tensor)</code></li>
</ul>
<p>Note: These operations are allowed when executing eagerly.</p>
<p>When encountering this error, make sure that the function is either decorated
with <code>@tf.function</code>, or called from another function decorated in this way. Also
look at the console and logging output for conversion warnings (see the section
above).</p>
<h3>"OperatorNotAllowedInGraphError: iterating over <code>tf.Tensor</code>"</h3>
<p>This exception is raised whenever you try to iterate over a <code>tf.Tensor</code>,
in a context where eager execution is not active. The exception is only raised
when graph execution is active, for example inside a <code>@tf.function</code> with
AutoGraph turned off. It can be caused by using a <code>tf.Tensor</code> value as:</p>
<ul>
<li>the iterated of a <code>for</code> statement: <code>for i in tensor:</code></li>
<li>the argument to the <code>iter</code> built-in: <code>iter(tensor)</code></li>
</ul>
<p>Note: These operations are allowed when executing eagerly.</p>
<p>This exception is similar to the previous example, and has similar causes and
remedies.</p>
<h3>"InaccessibleTensorError: The tensor <code>&lt;name&gt;</code> is defined in another function or code block"</h3>
<p>This exception is common to code which attempts to obtain values calculated
within a <code>tf.cond</code>, <code>tf.while_loop</code>, or another <code>@tf.function</code> without using
functional style or through mutable collections. See
<a href="https://www.tensorflow.org/guide/function#all_outputs_of_a_tffunction_must_be_return_values">Capturing External Symbolic Tensors</a>
and <a href="limitations.md">Limitations</a> for more details.</p>
<h3>"StagingError: in converted code"</h3>
<p>This exception is used by AutoGraph to wrap exceptions with custom constructors
that it cannot re-raise with the original type. See
<a href="error_handling.md">Error handling</a> for more details. If your code uses custom
exceptions, expect them to be wrapped by this exception.</p>
<h3>"Unable to identify source code of lambda function"</h3>
<p>This error usually appears in the context of a conversion warning. It indicates
that a lambda function could not be parsed (see <a href="limitations.md">Limitations</a>).</p>
<p>This type of errors can usually be avoided by creating lambda functions in
separate simple assignments, for example:</p>
<p><code>l = lambda &lt;args&gt;: &lt;body&gt;</code></p>